### PR TITLE
migrating water integration into one repository

### PR DESCRIPTION
**Pull request by WendelHime**
This PR migrate WATER integration into this repository because there's some shared usage between flashlight and http-proxy like the dependency of WASM downloaders. Most of the code is the same and whenever there's a difference I've left comments.

### CODE REVIEW

File: dialer/dialer.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import (
+	"context"
+	"log/slog"
+
+	"github.com/getlantern/golog"
+	"github.com/getlantern/lantern-water/logger"
+	"github.com/refraction-networking/water"
+)
+
+type DialerParameters struct {
```
Comment (WendelHime) [inline dialer/dialer.go:15]: This is new and didn't exist previously at flashlight

File: listener/listener.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	Logger golog.Logger
+	// Transport represents the protocol, version or whatever detail that will
+	// be used at local logs to help understanding which WASM file is being used
+	Transport string
+
+	// Address represents the address used by the listener
+	Address string
+	// WASM must contain the WASM data used by the WATER listener
+	WASM []byte
+}
```
Comment (WendelHime) [inline listener/listener.go]: This part also didn't exist before at http-proxy

**Review by hwh33 — APPROVED** [review]
This all looks great @WendelHime!  Could you just make sure all of the exported definitions have documentation?  That includes package-level documentation.  Feel free to merge after that!

File: dialer/dialer.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+// NewDialer creates a new water dialer with the given parameters.
+func NewDialer(ctx context.Context, params DialerParameters) (water.Dialer, error) {
+	cfg := &water.Config{
+		TransportModuleBin: params.WASM,
+	}
+
+	if params.Logger != nil {
+		cfg.OverrideLogger = slog.New(logger.NewLogHandler(params.Logger, params.Transport))
+	}
```
Comment (WendelHime) [inline dialer/dialer.go:32]: This is also new, now the Logger is optional otherwise it'll use the default water logger

**Pull request merged**